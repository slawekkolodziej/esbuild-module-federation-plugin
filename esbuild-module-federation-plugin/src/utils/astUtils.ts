import babelParser from "@babel/parser";
import generate from "@babel/generator";
import t, {
  CallExpression,
  VariableDeclaration,
  ImportDeclaration,
  VariableDeclarator,
} from "@babel/types";

type ImportSpecifiers = {
  defaultImport: string | null;
  namedImports: Array<{ local: string; imported: string }>;
};

type MaybeVariableDeclarator = VariableDeclarator | null;

export function codeToAst(code: string) {
  return babelParser.parse(code, {
    sourceType: "module",
    plugins: ["jsx"],
  });
}

export function astToCode(ast): string {
  return generate(ast).code;
}

export function replaceNode(node, newNode) {
  const currentKeys = Object.keys(node);
  const newKeys = Object.keys(newNode);

  Object.assign(node, newNode);

  currentKeys
    .filter((key) => !newKeys.includes(key))
    .forEach((key) => {
      delete node[key];
    });
}

export function convertImportDeclarations(
  node: ImportDeclaration,
  callExpression: CallExpression
): VariableDeclaration {
  const { defaultImport, namedImports } =
    node.specifiers.reduce<ImportSpecifiers>(
      (acc, specifier) => {
        if (t.isImportDefaultSpecifier(specifier)) {
          acc.defaultImport = specifier.local.name;
        } else if (t.isImportSpecifier(specifier)) {
          acc.namedImports.push({
            imported: t.isStringLiteral(specifier.imported)
              ? specifier.imported.value
              : specifier.imported.name,
            local: specifier.local.name,
          });
        }
        return acc;
      },
      { defaultImport: null, namedImports: [] }
    );

  const newNode = t.variableDeclaration(
    "const",
    [
      createDeclaratorForDefaultImport(defaultImport, callExpression),
      createDeclaratorForNamedImports(
        namedImports,
        defaultImport,
        callExpression
      ),
    ].filter(isVariableDeclarator)
  );

  return newNode;
}

function isVariableDeclarator(
  decl: MaybeVariableDeclarator
): decl is VariableDeclarator {
  return decl !== null;
}

function createDeclaratorForDefaultImport(
  defaultImportName: ImportSpecifiers["defaultImport"],
  callExpression: CallExpression
): MaybeVariableDeclarator {
  if (!defaultImportName) {
    return null;
  }

  return t.variableDeclarator(t.identifier(defaultImportName), callExpression);
}

function createDeclaratorForNamedImports(
  namedImports: ImportSpecifiers["namedImports"],
  defaultImportName: ImportSpecifiers["defaultImport"],
  callExpression: CallExpression
): MaybeVariableDeclarator {
  if (namedImports.length === 0) {
    return null;
  }

  return t.variableDeclarator(
    t.objectPattern(
      namedImports.map(({ imported, local }) =>
        t.objectProperty(
          t.identifier(imported),
          t.identifier(local),
          false,
          imported === local
        )
      )
    ),
    defaultImportName ? t.identifier(defaultImportName) : callExpression
  );
}
