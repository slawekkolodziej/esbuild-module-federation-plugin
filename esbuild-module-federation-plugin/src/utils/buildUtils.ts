export function getRelativeChunkPath(build) {
  return (
    (build.initialOptions.chunkNames ?? "[name]-[hash]")
      .split("/")
      .slice(0, -1)
      .map(() => "..")
      .join("/") || "."
  );
}

export function generateUniqueIdentifier(code, prefix = "var") {
  let i = 0;
  let varName = `__${prefix}${i}`;

  while (code.indexOf(varName) > -1) {
    varName = `${prefix}${i}`;
    i++;
  }

  return varName;
}
