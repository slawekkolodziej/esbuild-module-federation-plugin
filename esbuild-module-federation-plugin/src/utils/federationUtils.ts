function getVersion(shared) {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  return require(require.resolve(`${shared}/package.json`)).version;
}

function normalizeShared(shared) {
  shared = shared || [];

  if (Array.isArray(shared)) {
    return shared.reduce((acc, shared) => {
      return {
        ...acc,
        [shared]: {
          shareKey: shared,
          shareScope: ["default"],
          version: getVersion(shared),
        },
      };
    }, {});
  }

  return Object.entries(shared).reduce((acc, [shared, sharedConfig]) => {
    let shareScope = ["default"];
    if (Array.isArray(sharedConfig.shareScope)) {
      shareScope = sharedConfig.shareScope;
    } else if (typeof sharedConfig.shareScope === "string") {
      shareScope = [sharedConfig.shareScope];
    }

    return {
      ...acc,
      [shared]: {
        shareKey: sharedConfig.shareKey || shared,
        shareScope,
        version: sharedConfig.version || getVersion(shared),
      },
    };
  }, {});
}

function normalizeModuleName(moduleName) {
  if (moduleName[0] === "@") {
    const [modScope, modName, ...modPath] = moduleName.split("/");

    return [[modScope, modName].join("/"), `./${modPath.join("/")}`];
  }

  const [modName, ...modPath] = moduleName.split("/");

  return [modName, `./${modPath.join("/")}`];
}

function normalizeRemotes(remotes = {}) {
  const remoteAddressWithGlobalRe = /^(?:([a-z-_][a-z0-9-_]+)@)?(.+)$/i;

  return Object.entries(remotes).reduce((acc, [remote, remoteConfig]) => {
    const config = {};

    if (typeof remoteConfig === "string") {
      const [, global, src] = remoteConfig.match(remoteAddressWithGlobalRe);

      config.type = "var";
      config.global = global;
      config.src = src;
    }

    return {
      ...acc,
      [remote]: {
        src: config.src || remoteConfig.src || "",
        type: config.type || remoteConfig.type || "var",
        global: config.global || remoteConfig.global || remote,
      },
    };
  }, {});
}

export { normalizeModuleName, normalizeRemotes, normalizeShared, getVersion };
