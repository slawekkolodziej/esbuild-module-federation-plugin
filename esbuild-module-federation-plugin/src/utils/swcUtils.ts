import t from "@babel/types";

function createSpan() {
  return {
    start: 0,
    end: 0,
    ctxt: 0,
  };
}

function babelToSwc(fn) {
  return (...args) => {
    const node = fn(...args);

    return {
      ...node,
      span: createSpan(),
    };
  };
}

const arrowFunctionExpression = babelToSwc(t.arrowFunctionExpression);
const blockStatement = babelToSwc(t.blockStatement);
const callExpression = babelToSwc(t.callExpression);
const identifier = (name) => ({
  type: "Identifier",
  value: name,
  span: createSpan(),
});
const expression = (expression) => ({
  type: "Expression",
  expression,
  spread: null,
});
const ifStatement = babelToSwc(t.ifStatement);
const importDeclaration = babelToSwc(t.importDeclaration);
const importSpecifier = babelToSwc(t.importSpecifier);
const memberExpression = babelToSwc(t.memberExpression);
const objectProperty = babelToSwc(t.objectProperty);
const regExpLiteral = babelToSwc(t.regExpLiteral);
const returnStatement = babelToSwc(t.returnStatement);
const stringLiteral = babelToSwc(t.stringLiteral);
const variableDeclaration = babelToSwc(t.variableDeclaration);
const variableDeclarator = babelToSwc(t.variableDeclarator);

const isExportNamedDeclaration = t.isExportNamedDeclaration;
const isExpressionStatement = t.isExpressionStatement;
const isImportDefaultSpecifier = t.isImportDefaultSpecifier;
const isVariableDeclaration = t.isVariableDeclaration;

const objectPattern = (properties) => ({
  type: "ObjectPattern",
  properties,
  span: createSpan(),
  optional: false,
});

const assignmentPatternProperty = (key) => ({
  type: "AssignmentPatternProperty",
  key,
  value: null,
  span: createSpan(),
  optional: false,
});

const keyValuePatternProperty = (key, value) => ({
  type: "KeyValuePatternProperty",
  key,
  value,
  span: createSpan(),
});

export {
  arrowFunctionExpression,
  blockStatement,
  callExpression,
  expression,
  identifier,
  ifStatement,
  importDeclaration,
  importSpecifier,
  memberExpression,
  objectPattern,
  objectProperty,
  assignmentPatternProperty,
  keyValuePatternProperty,
  regExpLiteral,
  returnStatement,
  stringLiteral,
  variableDeclaration,
  variableDeclarator,
  isExportNamedDeclaration,
  isExpressionStatement,
  isImportDefaultSpecifier,
  isVariableDeclaration,
};
