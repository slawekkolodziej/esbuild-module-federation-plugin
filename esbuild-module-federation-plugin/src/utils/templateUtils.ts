export function templateToFunction(template: string) {
  return new Function("a,b,c,d,e", `return (${template})(a,b,c,d,e);`);
}
