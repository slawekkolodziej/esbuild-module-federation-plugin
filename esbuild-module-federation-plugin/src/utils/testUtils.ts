import path from "path";
import vm from "vm";
import { promisify } from "util";
import rimraf from "rimraf";
import esbuild from "esbuild";
import babelParser from "@babel/parser";
import generate from "@babel/generator";

export const emptyDir = promisify(rimraf);

function createSyntheticModule(moduleMock, context) {
  const exportNames = Object.keys(moduleMock);

  return new vm.SyntheticModule(
    exportNames,
    function () {
      exportNames.forEach((exportName) => {
        this.setExport(exportName, moduleMock[exportName]);
      });
    },
    { context }
  );
}

function createModule(moduleMock, context, importModuleDynamically) {
  return new vm.SourceTextModule(moduleMock, {
    importModuleDynamically,
    context,
  });
}

function linkerFactory(modules, context) {
  const linker = (dep) => {
    const moduleMock = modules[dep];

    if (typeof moduleMock === "string") {
      return createModule(moduleMock, context, dynamicLoader);
    }

    return createSyntheticModule(moduleMock, context);
  };

  const dynamicLoader = async (moduleName) => {
    const esModule = linker(moduleName);

    await esModule.link(linker);
    await esModule.evaluate();

    return esModule;
  };

  return {
    dynamicLoader,
    linker,
  };
}

export async function mockModule(code, modules) {
  const context = vm.createContext({
    require: (mod) => modules[mod],
    globalThis: {},
  });
  const { dynamicLoader, linker } = linkerFactory(modules, context);
  const moduleMock = createModule(code, context, dynamicLoader);
  await moduleMock.link(linker);
  await moduleMock.evaluate();

  return [moduleMock, context.globalThis];
}

const WHITE_SPACE_RE = /^(\s+)/;

export function trimIndent(str = '') {
  const lines = str.trim().split("\n");

  return lines.map((line) => line.replace(WHITE_SPACE_RE, "")).join("\n");
}

export async function buildFixture(fixtureName, options) {
  const src = path.resolve(__dirname, "../../fixtures", fixtureName, "src");
  const outdir = path.resolve(src, "../out");

  await emptyDir(outdir);

  await esbuild.build({
    ...options,
    absWorkingDir: src,
    outdir,
  });

  return outdir;
}
