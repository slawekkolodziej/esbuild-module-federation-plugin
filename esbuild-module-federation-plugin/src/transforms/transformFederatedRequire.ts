import { dirname, join } from "path";
import { readFile, writeFile, rm } from "fs/promises";
import t from "@babel/types";
import { SHARED_SCOPE_MODULE_NAME, FEDERATED_MODULE_RE_STR } from "../const";
import { astToCode, codeToAst } from "../utils/astUtils";
import { generateUniqueIdentifier } from "../utils/buildUtils";
import { createSharedScopeImport } from "./transformFederatedEsmImports";

type TransformFederatedRequireRetVal = {
  requireMockCode: string;
  requireMockChunk: string;
  requireNamedExport: string;
};

export async function transformFederatedRequireCjs(
  filePath: string,
  relativeChunkPath = "."
) {
  // const buildRoot = dirname(requireMockPath);
  const code = await readFile(filePath, "utf-8");
  const originalRequireFn = generateUniqueIdentifier(code, "originalRequire");
  const federatedModuleRe = generateUniqueIdentifier(code, "federatedModuleRe");

  const requireMock = /* js */ `
    var Module = require('module');
    var { getModule } = require('${relativeChunkPath}/${SHARED_SCOPE_MODULE_NAME}.js')
    var ${originalRequireFn} = Module.prototype.require;
    var ${federatedModuleRe} = new RegExp(
      '${FEDERATED_MODULE_RE_STR.replace(/\//g, "\\/")})'
    );

    Module.prototype.require = function(mod) {
      if (mod.match(${federatedModuleRe})) {
        return getModule(mod);
      }
      return ${originalRequireFn}.apply(this, arguments);
    };
  `;

  return await writeFile(filePath, [requireMock, code].join("\n"));
}

export async function transformFederatedRequireEsm(
  requireMockPath: string,
  relativeChunkPath = "."
): Promise<TransformFederatedRequireRetVal> {
  const buildRoot = dirname(requireMockPath);
  const code = await readFile(requireMockPath, "utf-8");
  const requireMockChunkDetails = locateRequireChunk(codeToAst(code));
  const requireChunkPath = join(
    buildRoot,
    requireMockChunkDetails.requireMockChunk
  );
  const [requireMockCode] = await Promise.all([
    readFile(requireChunkPath, "utf-8"),
    rm(requireMockPath),
  ]);
  const finalAst = alterGlobalRequire(
    codeToAst(requireMockCode),
    requireMockChunkDetails.requireNamedExport,
    relativeChunkPath
  );
  const updatedRequireMockCode = astToCode(finalAst);

  await writeFile(requireChunkPath, updatedRequireMockCode);

  return {
    requireMockCode: updatedRequireMockCode,
    ...requireMockChunkDetails,
  };
}

type LocateRequireChunkRetVal = {
  requireMockChunk: string;
  requireNamedExport: string;
};

export function locateRequireChunk(ast): LocateRequireChunkRetVal {
  const iife = ast.program.body.find(
    (node) =>
      t.isExpressionStatement(node) &&
      t.isCallExpression(node.expression) &&
      node.expression.arguments.length > 0
  );
  const iifeArguments = iife.expression.arguments;
  const localRequireName = iifeArguments[0].name;

  if (!localRequireName) {
    throw new Error('Could not found local "require" mock!');
  }

  const requireMockImportDecl = ast.program.body.find(
    (node) =>
      node.type === "ImportDeclaration" &&
      node.specifiers.some(
        (specifier) => specifier.local.name === localRequireName
      )
  );
  const requireMockChunk = requireMockImportDecl.source.value;

  if (!requireMockChunk) {
    throw new Error('Could not found "require" mock chunk!');
  }

  const requireNamedExport = requireMockImportDecl.specifiers[0].imported.name;

  return {
    requireMockChunk,
    requireNamedExport,
  };
}

function alterGlobalRequire(ast, namedExport, relativeChunkPath) {
  const namedExportDecl = ast.program.body.find((node) =>
    t.isExportNamedDeclaration(node)
  );

  if (!namedExportDecl) {
    throw new Error("Could not find named exports node.");
  }

  const requireMockExportSpecifier = namedExportDecl.specifiers.find(
    (specifier) => specifier.exported.name === namedExport
  );

  if (!requireMockExportSpecifier) {
    throw new Error("Could not find named export for require mock.");
  }

  const localName = requireMockExportSpecifier.local.name;
  const isRequireDecl = (node) => node.id.name === localName;
  const requireMockDeclGroup = ast.program.body.find(
    (node) =>
      t.isVariableDeclaration(node) && node.declarations.some(isRequireDecl)
  );
  const requireMockDecl = requireMockDeclGroup.declarations.find(isRequireDecl);

  if (!requireMockDecl) {
    throw new Error("Could not find require mock declaration.");
  }

  // The code below should generate the following syntax:
  // (req => mod => {
  //   if (mod.matches(FEDERATED_MODULE_RE)) {
  //       return getModule(mod);
  //   }
  //   return req(mod);
  // })(/* original syntax for __require would go here */)
  requireMockDecl.init = t.callExpression(
    t.arrowFunctionExpression(
      [t.identifier("req")],
      t.arrowFunctionExpression(
        [t.identifier("mod")],
        t.blockStatement([
          t.ifStatement(
            t.callExpression(
              t.memberExpression(t.identifier("mod"), t.identifier("match")),
              [t.regExpLiteral(FEDERATED_MODULE_RE_STR.replace(/\//g, "\\/"))]
            ),
            t.blockStatement([
              t.returnStatement(
                t.callExpression(t.identifier("getModule"), [
                  t.identifier("mod"),
                ])
              ),
            ])
          ),
          t.returnStatement(
            t.callExpression(t.identifier("req"), [t.identifier("mod")])
          ),
        ])
      )
    ),
    [requireMockDecl.init]
  );

  ast.program.body.unshift(
    createSharedScopeImport(relativeChunkPath, {
      getModule: true,
      getModuleAsync: false,
    })
  );

  return ast;
}
