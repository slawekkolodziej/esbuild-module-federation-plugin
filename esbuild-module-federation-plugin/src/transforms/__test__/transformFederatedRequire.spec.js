import path from "path";
import { transformFederatedRequire } from "../transformFederatedRequire";
import { buildFixture } from "../../utils/testUtils";

const esbuildOptions = {
  entryPoints: {
    app1: "app1.jsx",
    app2: "app2.jsx",
  },
  bundle: true,
  chunkNames: "chunks/[name]-[hash]",
  format: "esm",
  minify: false,
  splitting: true,
  write: true,
};

describe("transformFederatedRequire", () => {
  it("properly overrides __require function", async () => {
    const out = await buildFixture("transformFederatedRequire", esbuildOptions);
    const { remoteEntryCode, requireMockCode } =
      await transformFederatedRequire(path.join(out, "app1.js"));

    expect(remoteEntryCode).toMatchSnapshot();
    expect(requireMockCode).toMatchSnapshot();
  });

  it("properly overrides __require function in minified code", async () => {
    const out = await buildFixture(
      "transformFederatedRequire",
      {
        ...esbuildOptions,
        minify: true,
      },
      "out-min"
    );
    const { remoteEntryCode, requireMockCode } =
      await transformFederatedRequire(path.join(out, "app1.js"));

    expect(remoteEntryCode).toMatchSnapshot();
    expect(requireMockCode).toMatchSnapshot();
  });
});
