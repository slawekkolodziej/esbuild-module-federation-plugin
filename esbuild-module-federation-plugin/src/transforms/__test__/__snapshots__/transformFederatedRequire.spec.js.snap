// Vitest Snapshot v1

exports[`transformFederatedRequire > properly overrides __require function 1`] = `
"// app1.jsx
(function (self) {})(globalThis);"
`;

exports[`transformFederatedRequire > properly overrides __require function 2`] = `
"import { getModule } from \\"./federation-shared.js\\";

var __require = (req => mod => {
  if (mod.match(/@runtime\\\\/federation\\\\/(remote|shared)\\\\/(.+)/)) {
    return getModule(mod);
  }

  return req(mod);
})(x => {
  if (typeof require !== \\"undefined\\") return require(x);
  throw new Error('Dynamic require of \\"' + x + '\\" is not supported');
}); // app2.jsx


(function (self) {})(globalThis, __require);

function App() {}

export { __require, App };"
`;

exports[`transformFederatedRequire > properly overrides __require function in minified code 1`] = `"(function (i) {})(globalThis);"`;

exports[`transformFederatedRequire > properly overrides __require function in minified code 2`] = `
"import { getModule } from \\"./federation-shared.js\\";

var f = (req => mod => {
  if (mod.match(/@runtime\\\\/federation\\\\/(remote|shared)\\\\/(.+)/)) {
    return getModule(mod);
  }

  return req(mod);
})(e => {
  if (typeof require != \\"undefined\\") return require(e);
  throw new Error('Dynamic require of \\"' + e + '\\" is not supported');
});

(function (e) {})(globalThis, f);

function i() {}

export { f as a, i as b };"
`;
